lib/datalog/roletester/src/role_tester.rs

```diff
@@ -0,0 +1,136 @@
+}
+
+#[no_mangle]
+pub extern "C" fn process_access(input: *mut c_uchar, output: *mut c_uchar, input_len: size_t, output_len: size_t) -> i32 {
+    let mut runtime = Crepe::new();
+    let b = unsafe { slice::from_raw_parts_mut(input, input_len) };
+    let r = match types::Facts::decode(BytesMut::from(&b[..])) {
+        Ok(b) => b,
+        Err(_e) => return Errors::InvalidInputError as i32
+    };
+
+    for pred in &r.predicates {
+        if pred.name == types::facts::PredicateType::Hasrole as i32 {
+            runtime.extend(&[HasRole(pred.atoms[0], pred.atoms[1])]);
+        } else if pred.name == types::facts::PredicateType::Hastrait as i32 {
+            runtime.extend(&[HasTrait(pred.atoms[0], pred.atoms[1], pred.atoms[2])]);
+        } else if pred.name == types::facts::PredicateType::Roleallowslogin as i32 {
+            runtime.extend(&[RoleAllowsLogin(pred.atoms[0], pred.atoms[1])]);
+        } else if pred.name == types::facts::PredicateType::Roledenieslogin as i32 {
+            runtime.extend(&[RoleDeniesLogin(pred.atoms[0], pred.atoms[1])]);
+        } else if pred.name == types::facts::PredicateType::Roleallowsnodelabel as i32 {
+            runtime.extend(&[RoleAllowsNodeLabel(pred.atoms[0], pred.atoms[1], pred.atoms[2])]);
+        } else if pred.name == types::facts::PredicateType::Roledeniesnodelabel as i32 {
+            runtime.extend(&[RoleDeniesNodeLabel(pred.atoms[0], pred.atoms[1], pred.atoms[2])]);
+        } else if pred.name == types::facts::PredicateType::Nodehaslabel as i32 {
+            runtime.extend(&[NodeHasLabel(pred.atoms[0], pred.atoms[1], pred.atoms[2])]);
+        } else {
+            return Errors::InvalidInputError as i32;
+        }
+    }
+
+    let (accesses, deny_accesses, deny_logins) = runtime.run();
+    let mut predicates = vec![];
+    predicates.extend::<Vec<_>>(accesses.into_iter().map(|HasAccess(a, b, c, d)| types::facts::Predicate{
+        name: types::facts::PredicateType::Hasaccess as i32,
+        atoms: vec![a, b, c, d],
+    }).collect());
+    predicates.extend::<Vec<_>>(deny_accesses.into_iter().map(|DenyAccess(a, b, c, d)| types::facts::Predicate{
+        name: types::facts::PredicateType::Denyaccess as i32,
+        atoms: vec![a, b, c, d],
+    }).collect());
+    predicates.extend::<Vec<_>>(deny_logins.into_iter().map(|DenyLogins(a, b, c)| types::facts::Predicate{
+        name: types::facts::PredicateType::Denylogins as i32,
+        atoms: vec![a, b, c],
+    }).collect());
+
+    let idb = types::Facts {
+        predicates: predicates
+    };
+
+    let mut buf = Vec::with_capacity(idb.encoded_len());
+    match idb.encode(&mut buf) {
+        Ok(_) => (),
+        Err(_e) => return Errors::InvalidOutputError as i32,
+    };
+    unsafe {
+        if buf.len() == 0 || buf.len() > output_len {
+            return buf.len() as i32
+        }
+        ptr::copy_nonoverlapping(&(*buf)[0], output, buf.len());
```

Comment:
This needs a detailed annotating comment explaining why it's safe and what invariants it relies on.